File: frontend/src/auth/roleMapping.ts
import type { UserRole } from "./permissions";

// =====================================================================
// Keycloak realm rollerini uygulama içi rollerle eşliyoruz.
// ---------------------------------------------------------------------
// Token içindeki realm_access.roles dizisi, Keycloak realm'inde
// TANIMLI olan rol adlarını içerir. Bu isimler admin konsolunda
// belirlenir — sen admin konsoluna giremediğin için gerçek isimleri
// şu an bilmiyoruz. Aşağıdaki eşleme bir TAHMİNDİR/iskeledir.
//
// Gerçek realm rol adları netleşince (Keycloak'ı yöneten ekipten
// öğrenerek) SADECE bu dosyadaki ROLE_MAP güncellenecek — başka
// hiçbir dosyaya dokunmaya gerek kalmayacak.
// =====================================================================
const ROLE_MAP: Record<string, UserRole> = {
  yonetici: "Yönetici",
  admin: "Yönetici",
  analist: "Analist",
  supervisor: "Supervisor",
};

/**
 * Keycloak'tan gelen realm rollerinden, uygulamanın anladığı TEK bir
 * role karar verir. Birden fazla eşleşen rol varsa ilk bulunanı
 * kullanır (birden fazla rolün aynı anda atanması normalde beklenmez).
 *
 * Hiçbir rol eşleşmezse `null` döner; çağıran taraf bunu "bilinmeyen rol"
 * olarak ele alıp EN KISITLI davranışı uygulamalıdır (bkz. auth/permissions.ts
 * -> isKnownRole, "izin yok" varsayımı güvenli taraftır).
 */
export function mapKeycloakRolesToAppRole(
  realmRoles: string[],
): UserRole | null {
  for (const rawRole of realmRoles) {
    const normalized = rawRole.trim().toLowerCase();
    if (normalized in ROLE_MAP) {
      return ROLE_MAP[normalized];
    }
  }
  return null;
}

File: backend/SesKayitlariApi/auth/RoleMapping.cs
namespace SesKayitlariApi.Auth;

/// <summary>
/// Keycloak realm rollerini uygulama içi UserRole'e eşler.
///
/// =====================================================================
/// ÖNEMLİ — BU DOSYA frontend/src/auth/roleMapping.ts İLE BİREBİR
/// TUTARLI OLMALI. İki taraf da aynı realm rol adlarını aynı UserRole
/// değerine eşlemeli; aksi halde bir kullanıcı frontend'de "Yönetici"
/// olarak görünürken backend'de farklı bir role (veya hiçbir role)
/// eşlenebilir — bu durum tutarsız/güvensiz yetkilendirmeye yol açar.
///
/// Aşağıdaki eşleme frontend'deki gibi bir TAHMİNDİR/iskeledir — gerçek
/// realm rol adları Keycloak'ı yöneten ekipten öğrenilince, BU DOSYA ve
/// frontend/src/auth/roleMapping.ts AYNI ANDA güncellenmeli. Rol adları
/// netleşene kadar, hiçbir gerçek kullanıcının doğru şekilde
/// yetkilendirilmeyeceğini unutma.
/// =====================================================================
/// </summary>
public static class RoleMapping
{
    // Anahtarlar küçük harfle tutulur; token'dan gelen rol adı
    // karşılaştırılmadan önce ToLowerInvariant() ile normalize edilir
    // (bkz. MapRealmRolesToAppRole). Bu, frontend'deki
    // rawRole.trim().toLowerCase() davranışıyla birebir aynıdır.
    private static readonly Dictionary<string, UserRole> RoleMap = new()
    {
        ["yonetici"] = UserRole.Yonetici,
        ["admin"] = UserRole.Yonetici,
        ["analist"] = UserRole.Analist,
        ["supervisor"] = UserRole.Supervisor,
    };

    /// <summary>
    /// Token'daki realm_access.roles dizisinden, uygulamanın anladığı
    /// TEK bir role karar verir. Birden fazla eşleşen rol varsa ilk
    /// bulunanı kullanır (frontend'deki mapKeycloakRolesToAppRole ile
    /// aynı davranış).
    ///
    /// Hiçbir rol eşleşmezse null döner; çağıran taraf bunu "bilinmeyen
    /// rol" olarak ele alıp EN KISITLI davranışı (yetki yok) uygulamalı.
    /// </summary>
    public static UserRole? MapRealmRolesToAppRole(IEnumerable<string> realmRoles)
    {
        foreach (var rawRole in realmRoles)
        {
            var normalized = rawRole.Trim().ToLowerInvariant();
            if (RoleMap.TryGetValue(normalized, out var appRole))
            {
                return appRole;
            }
        }

        return null;
    }
}